libff/algebra/fields/binary_field.hpp

```diff
@@ -0,0 +1,110 @@
+ * @author     This file is part of libff, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+#include <libff/algebra/field_utils/bigint.hpp>
+#include <vector>
+
+namespace libff {
+
+template<typename T>
+class BinaryField;
+
+/* The type parameter T is intended to be set to the child class
+ * when this class is extended. For example,
+ * class gf32 : public BinaryField<gf32> ...
+ */
+template<typename T>
+class BinaryField {
+public:
+    /* Functions unique to binary fields */
+
+    static const constexpr uint64_t modulus_;
+    static const constexpr uint64_t num_bits;
+
+    /** generator of gf2^n */
+    static T multiplicative_generator;
+
+    /** If extension field, returns the base field's characteristic. */
+    template<mp_size_t n>
+    static constexpr bigint<n> field_char() { return bigint<n>(2); }
+
+    /* Functions common to all finite fields */
+
+#ifdef PROFILE_OP_COUNTS // NOTE: op counts are affected when you exponentiate with ^
+    static long long add_cnt;
+    static long long sub_cnt;
+    static long long mul_cnt;
+    static long long sqr_cnt;
+    static long long inv_cnt;
+#endif
+
+    virtual T& operator+=(const T& other) = 0;
+    virtual T& operator-=(const T& other) = 0;
+    virtual T& operator*=(const T& other) = 0;
+    virtual T& operator^=(const unsigned long pow) = 0;
+    template<mp_size_t m>
+    virtual T& operator^=(const bigint<m> &pow) = 0;
+
+    virtual T& square() = 0;
+    virtual T& invert() = 0;
+
+    virtual T operator+(const T& other) const;
+    virtual T operator-(const T& other) const;
+    virtual T operator*(const T& other) const;
+    virtual T operator^(const unsigned long pow) const;
+    template<mp_size_t m>
+    virtual T operator^(const bigint<m> &pow) const;
+    virtual T operator-() const = 0;
+
+    virtual T squared() const;
+    virtual T inverse() const;
+    /** HAS TO BE A SQUARE (else does not terminate). */
+    virtual T sqrt() const = 0;
+
+    bool operator==(const T& other) const = 0;
+    bool operator!=(const T& other) const = 0;
+    bool is_zero() const = 0;
+
+    void print() const = 0;
+    /**
+     * Returns the constituent bits in 64 bit words, in little-endian order.
+     * Only the right-most size_in_bits() bits are used; other bits are 0.
+     */
+    std::vector<uint64_t> as_words() const = 0;
+
+    void randomize() = 0;
+    void clear() = 0;
+
+    // the following should be defined in child classes, but are static so they can't be inherited
+    static T zero();
+    static T one();
+    static T random_element();
+    /**
+     * Creates a field element from the given bits in 64 bit words, in little-endian order.
+     * Only the right-most size_in_bits() bits are used; other bits are ignored.
+     */
+    static T from_words(std::vector<uint64_t> words);
```

Comment:
can you move this to be next to `as_words`, and perhaps rename `as_words` to `to_words` for name consistency?

Reply:
I put it here because I was placing static functions together. Do you think it's better to move it?

Reply:
Yeah I think its better to move it

Reply:
Hmm I feel like I disagree, but eh it's not a big deal

Reply:
I decided to move as_words (now to_words) down, instead of moving to_words up